File: app/src/main/java/ru/nikitazhelonkin/coinbalance/data/api/exchange/BittrexApiClient.java
package ru.nikitazhelonkin.coinbalance.data.api.exchange;


import java.util.HashMap;

import io.reactivex.Single;
import ru.nikitazhelonkin.coinbalance.data.api.HttpErrorTransformer;
import ru.nikitazhelonkin.coinbalance.data.api.response.BittrexBalancesResponse;
import ru.nikitazhelonkin.coinbalance.utils.DigestUtil;

public class BittrexApiClient implements ExchangeApiClient {

    private BittrexApiService mApiService;

    public BittrexApiClient(BittrexApiService apiService) {
        mApiService = apiService;
    }

    @Override
    public Single<HashMap<String, Float>> getBalances(String apiKey, String apiSecret) {
        String timestamp = String.valueOf(System.currentTimeMillis());
        String contentHash = DigestUtil.sha512String("");
        String preSignString = timestamp + "https://api.bittrex.com/v3/balances" + "GET" + contentHash;
        String signature = DigestUtil.hmacString(preSignString, apiSecret, "HmacSHA512");
        return mApiService.balances(apiKey, timestamp, contentHash, signature)
                .map(BittrexBalancesResponse::getNonZeroBalances)
                .compose(new HttpErrorTransformer<>());
    }
}

File: app/src/main/java/ru/nikitazhelonkin/coinbalance/data/api/exchange/BittrexApiClient.kt
package ru.nikitazhelonkin.coinbalance.data.api.exchange


import io.reactivex.Single
import ru.nikitazhelonkin.coinbalance.data.api.HttpErrorTransformer
import ru.nikitazhelonkin.coinbalance.data.exception.ApiError
import ru.nikitazhelonkin.coinbalance.utils.DigestUtil
import java.util.*

class BittrexApiClient(private val apiService: BittrexApiService) : ExchangeApiClient {

    override fun getBalances(apiKey: String, apiSecret: String): Single<HashMap<String, Float>> {
        val nonce = System.currentTimeMillis().toString()
        val url = String.format("https://bittrex.com/api/v1.1/account/getbalances?apikey=%s&nonce=%s", apiKey, nonce)
        val signature = DigestUtil.hmacString(url, apiSecret, "HmacSHA512")
        return apiService.balances(apiKey, nonce, signature)
                .map { response ->
                    if (response.isSuccess) {
                        response.nonZeroBalances
                    } else {
                        throw ApiError(response.message)
                    }
                }
                .compose(HttpErrorTransformer())
    }
}
